auth0/src/main/java/com/auth0/android/authentication/storage/CryptoUtil.java

```diff
@@ -135,17 +142,55 @@ public CryptoUtil(@NonNull Context context, @NonNull Storage storage, @NonNull S
             generator.generateKeyPair();
 
             return getKeyEntryCompat(keyStore);
-        } catch (KeyStoreException | IOException | NoSuchProviderException | InvalidAlgorithmParameterException | NoSuchAlgorithmException | CertificateException e) {
-            Log.e(TAG, "An error occurred while trying to obtain the RSA KeyPair Entry from the Android KeyStore.", e);
-            throw new KeyException("An error occurred while trying to obtain the RSA KeyPair Entry from the Android KeyStore.", e);
-        } catch (UnrecoverableEntryException e) {
-            //Remove keys and Retry
-            Log.w(TAG, "RSA KeyPair was deemed unrecoverable. Deleting the existing entry and trying again.");
+        } catch (CertificateException | InvalidAlgorithmParameterException | NoSuchProviderException | NoSuchAlgorithmException | KeyStoreException e) {
+            /*
+             * This exceptions are safe to be ignored:
+             *
+             * - CertificateException:
+             *      Thrown when certificate has expired (25 years..) or couldn't be loaded
+             * - KeyStoreException:
+             * - NoSuchProviderException:
+             *      Thrown when "AndroidKeyStore" is not available. Was introduced on API 18.
+             * - NoSuchAlgorithmException:
+             *      Thrown when "RSA" algorithm is not available. Was introduced on API 18.
+             * - InvalidAlgorithmParameterException:
+             *      Thrown if Key Size is other than 512, 768, 1024, 2048, 3072, 4096
+             *      or if Padding is other than RSA/ECB/PKCS1Padding, introduced on API 18
+             *      or if Block Mode is other than ECB
+             *
+             * However if any of this exceptions happens to be thrown (OEMs often change their Android distribution source code),
+             * all the checks performed in this class wouldn't matter and the device would not be compatible at all with it.
+             *
+             * Read more in https://developer.android.com/training/articles/keystore#SupportedAlgorithms
+             */
+            Log.e(TAG, "The device can't generate a new RSA Key pair.", e);
+            throw new IncompatibleDeviceException(e);
+        } catch (IOException | UnrecoverableEntryException e) {
+            /*
+             * Any of this exceptions mean the old key pair is somehow corrupted.
+             * We can delete it and let the user retry the operation.
+             *
+             * - IOException:
+             *      Thrown when there is an I/O or format problem with the keystore data.
+             * - UnrecoverableEntryException:
+             *      Thrown when the key cannot be recovered. Probably because it was invalidated by a Lock Screen change.
+             */
             deleteKeys();
-            return getRSAKeyEntry();
+            throw new CryptoException("The existing RSA key pair could not be recovered and has been deleted. " +
```

Comment:
This is the only 2 exceptions scenario where I consider a retry could fix any issue, since I'm deleting any corrupted existing keys if this happens... next time it's called, it will recreate them as explained in the comments.